### PR TITLE
Provide support for `x-goog-api-client` header

### LINKED ISSUE #35: Support `x-goog-api-client` in go
Using this document to get opinions before a PR:

Background
=========
We plan to support an `X-Goog-Api-Client` header that allows us to track the version information of the code generator and GAX libraries for each client call.

The format is `<clientlib>/<semver> gapic/<semver> gax/<semver> language/<semver>`
For example, in python: `gcloud-python/<gcloud-python-version> gapic/0.1.0 gax/0.5.0 python/2.7`

Proposal
=======
In order to support the header, we need to be able to pass optional `app_name` and `app_version` strings to the client struct on initialization.

The method signature for `NewClient` is currently:
`func NewClient(ctx context.Context, opts ...cloud.ClientOption) (*Client, error)`

We propose instead:
`func NewClient(ctx context.Context, opts ...gax.ClientOption) (*Client, error)`
where the interface `gax.ClientOption` is provided by the following:

within `gax-golang`
-> `WithCloudClientOption(cloud.ClientOption) ClientOption`
-> `WithAppVersion(appVersion string) ClientOption`
-> `WithAppName(appName string) ClientOption`

This design also allows us to implement client level options specific to the gapic surface:
-> `WithCallOptionsOverride(map[string][]gax.CallOption) ClientOption`
-> `WithTimeout(duration time.Duration) ClientOption`

Now, within `NewClient` we can generate the header string as a private global variable and provide it to all calls via `grpc.CallOption`.

GAX will also support a global `Version string` that contains a semantic version tag. It's a workaround for go's package versioning limitations, and not perfect, but hopefully sufficient for now.
@tbetbetbe's suggestion was to use a script to generate a go file with a version tag (maybe the commit hash, or github release version).

### PR DESCRIPTION
**Pull request by saicheems**
Also refactors the `ClientSettings` implementation.

Resolves #35

### CODE REVIEW

**Comment by saicheems** [comment]
@jba 
@pongad 
@jmuk 
@vchudnov-g 

PTAL: This change modifies the way default client settings/call settings are provided to reduce some verbosity. Also implements the client header.

Relies on changes in GAX that haven't been pushed, but are fairly innocuous - mostly making private structs public.

File: vgen/src/test/java/io/gapi/vgen/testdata/go_library.baseline [hunk 1]
```diff
@@ -56,41 +57,98 @@ import (
...
+        AppVersion: gax.Version(),
+        Endpoint: "library-example.googleapis.com:443",
+        Scopes: []string{
+            "https://www.googleapis.com/auth/library",
+            "https://www.googleapis.com/auth/cloud-platform",
+        },
+        CallSettings: map[string][]gax.CallOption{
+            "CreateShelf": []gax.CallOption{
+                gax.WithRetryCodes(nonIdempotentRetryCodes()),
+                defaultRetryOptions()...
```
Comment (jmuk) [inline vgen/src/test/java/io/gapi/vgen/testdata/go_library.baseline]: I think this is invalid syntactically. Instead, probably
```
append([]gax.CallOption(gax.WithRetryCodes(nonIdempotentRetryCodes()), defaultRetryOptions()...)
```
?

See the compile errors of https://play.golang.org/p/HptLq_gC7j FYI.

Comment (saicheems) [inline vgen/src/test/java/io/gapi/vgen/testdata/go_library.baseline]: You're right - good catch
Fixed.

**Comment by jmuk** [comment]
LGTM

**Comment by pongad** [comment]
lgtm

**Pull request merged**